BukkitObjectStream Wrapper.map null?

Hey, I have those two methods.

```java
    /**
     * Encode item stack.
     *
     * @param item the item
     * @return the string
     */
    @NotNull
    public static String encodeBase64(@NotNull ItemStack item) throws IOException {
        try(ByteArrayOutputStream str = new ByteArrayOutputStream();
            BukkitObjectOutputStream data = new BukkitObjectOutputStream(str)) {
            data.writeObject(item);

            return Base64.getEncoder().encodeToString(str.toByteArray());
        }
    }

    /**
     * Decode item stack.
     *
     * @param base64 the base 64
     * @return the item stack
     */
    @NotNull
    public static ItemStack decodeBase64(@NotNull String base64) throws IOException, ClassNotFoundException {
        try(ByteArrayInputStream str = new ByteArrayInputStream(Base64.getDecoder().decode(base64));
            BukkitObjectInputStream data = new BukkitObjectInputStream(str)) {

            return (ItemStack) data.readObject(); // Error is here in JUnit 4 Test-Case
        }
    }
```

The first one is accepted by MockBukkit the second one throws an exception: 

[00:06:39 SEVERE]: Could not call method 'public static org.bukkit.inventory.ItemStack org.bukkit.inventory.ItemStack.deserialize(java.util.Map)' of class org.bukkit.inventory.ItemStack for deserialization

java.io.IOException: Failed to deserialize object

	at org.bukkit.util.io.BukkitObjectInputStream.newIOException(BukkitObjectInputStream.java:59)
	at org.bukkit.util.io.BukkitObjectInputStream.resolveObject(BukkitObjectInputStream.java:51)
	at java.base/java.io.ObjectInputStream.checkResolve(ObjectInputStream.java:1793)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1742)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:514)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:472)
	at de.rytrox.spicy.item.ItemStacks.decodeBase64(ItemStacks.java:71)
	at de.rytrox.spicy.item.ItemStacksTest.shouldDecodeBase64ToItemStack(ItemStacksTest.java:75)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.mockito.internal.runners.DefaultInternalRunner$1$1.evaluate(DefaultInternalRunner.java:55)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.mockito.internal.runners.DefaultInternalRunner$1.run(DefaultInternalRunner.java:100)
	at org.mockito.internal.runners.DefaultInternalRunner.run(DefaultInternalRunner.java:107)
	at org.mockito.internal.runners.StrictRunner.run(StrictRunner.java:41)
	at org.mockito.junit.MockitoJUnitRunner.run(MockitoJUnitRunner.java:163)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.NullPointerException: Cannot invoke "Object.getClass()" because the return value of "org.bukkit.configuration.serialization.ConfigurationSerialization.deserializeObject(java.util.Map)" is null
	at org.bukkit.util.io.BukkitObjectInputStream.resolveObject(BukkitObjectInputStream.java:49)
	... 40 more

--- a/src/main/java/be/seeseemelk/mockbukkit/MockUnsafeValues.java
+++ b/src/main/java/be/seeseemelk/mockbukkit/MockUnsafeValues.java
@@ -1,6 +1,7 @@
 package be.seeseemelk.mockbukkit;
 
 import com.destroystokyo.paper.util.VersionFetcher;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.Multimap;
 import io.papermc.paper.inventory.ItemRarity;
 import net.kyori.adventure.text.Component;
@@ -104,9 +105,9 @@ public class MockUnsafeValues implements UnsafeValues
 
 
 	@Override
-	public @NotNull Material toLegacy(@NotNull Material material)
+	public Material toLegacy(Material material)
 	{
-		if (material.isLegacy())
+		if (material == null || material.isLegacy())
 		{
 			return material;
 		}
@@ -116,20 +117,29 @@ public class MockUnsafeValues implements UnsafeValues
 	@Override
 	public Material fromLegacy(Material material)
 	{
-		return material;
+		if (material == null || !material.isLegacy())
+		{
+			return material;
+		}
+		throw new UnimplementedOperationException();
 	}
 
 	@Override
 	public Material fromLegacy(MaterialData material)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return fromLegacy(material, false);
 	}
 
 	@Override
-	public Material fromLegacy(MaterialData material, boolean itemPriority)
+	public Material fromLegacy(MaterialData materialData, boolean itemPriority)
 	{
-		// TODO Auto-generated method stub
+		// Paper will blindly call #getItemType even if materialData is null, so we might as well enforce that it isn't.
+		Preconditions.checkNotNull(materialData, "materialData cannot be null");
+		Material material = materialData.getItemType();
+		if (material == null || !material.isLegacy())
+		{
+			return material;
+		}
 		throw new UnimplementedOperationException();
 	}
 

--- a/src/test/java/be/seeseemelk/mockbukkit/UnsafeValuesTest.java
+++ b/src/test/java/be/seeseemelk/mockbukkit/UnsafeValuesTest.java
@@ -1,5 +1,7 @@
 package be.seeseemelk.mockbukkit;
 
+import org.bukkit.Material;
+import org.bukkit.material.MaterialData;
 import org.bukkit.plugin.InvalidDescriptionException;
 import org.bukkit.plugin.InvalidPluginException;
 import org.bukkit.plugin.PluginDescriptionFile;
@@ -11,9 +13,12 @@ import java.io.StringReader;
 import java.util.regex.Pattern;
 
 import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
+@SuppressWarnings("deprecation")
 class UnsafeValuesTest
 {
 
@@ -30,7 +35,7 @@ class UnsafeValuesTest
 	}
 
 	@AfterEach
-	void teardown()
+	void tearDown()
 	{
 		MockBukkit.unmock();
 	}
@@ -93,6 +98,62 @@ class UnsafeValuesTest
 			mockUnsafeValues.setMinimumApiVersion("1.15");
 			checkVersion("1.13");
 		});
+	}
+
+	@Test
+	void fromLegacy_Material_NotLegacy_ReturnsSame()
+	{
+		for (Material material : Material.values())
+		{
+			if (material.isLegacy())
+				continue;
+			assertEquals(material, mockUnsafeValues.fromLegacy(material));
+		}
+	}
+
+	@Test
+	void fromLegacy_Material_Legacy_Throws()
+	{
+		for (Material material : Material.values())
+		{
+			if (!material.isLegacy())
+				continue;
+			assertThrows(UnimplementedOperationException.class, () -> mockUnsafeValues.fromLegacy(material));
+		}
+	}
+
+	@Test
+	void fromLegacy_Material_Null_ReturnsNull()
+	{
+		assertNull(mockUnsafeValues.fromLegacy((Material) null));
+	}
+
+	@Test
+	void fromLegacy_MaterialData_NotLegacy_ReturnsSame()
+	{
+		for (Material material : Material.values())
+		{
+			if (material.isLegacy())
+				continue;
+			assertEquals(material, mockUnsafeValues.fromLegacy(new MaterialData(material)));
+		}
+	}
+
+	@Test
+	void fromLegacy_MaterialData_Legacy_Throws()
+	{
+		for (Material material : Material.values())
+		{
+			if (!material.isLegacy())
+				continue;
+			assertThrows(UnimplementedOperationException.class, () -> mockUnsafeValues.fromLegacy(new MaterialData(material)));
+		}
+	}
+
+	@Test
+	void fromLegacy_MaterialData_Null_ReturnsNull()
+	{
+		assertThrows(NullPointerException.class, () -> mockUnsafeValues.fromLegacy((MaterialData) null));
 	}
 
 }

--- a/src/test/java/be/seeseemelk/mockbukkit/UnsafeValuesTest.java
+++ b/src/test/java/be/seeseemelk/mockbukkit/UnsafeValuesTest.java
@@ -146,7 +146,8 @@ class UnsafeValuesTest
 		{
 			if (!material.isLegacy())
 				continue;
-			assertThrows(UnimplementedOperationException.class, () -> mockUnsafeValues.fromLegacy(new MaterialData(material)));
+			MaterialData materialData = new MaterialData(material);
+			assertThrows(UnimplementedOperationException.class, () -> mockUnsafeValues.fromLegacy(materialData));
 		}
 	}
 

